homeassistant/helpers/reload.py

```diff
@@ -138,15 +142,52 @@ async def _async_reconfig_platform(
     await asyncio.gather(*tasks)
 
 
+@overload
 async def async_integration_yaml_config(
     hass: HomeAssistant, integration_name: str
+) -> ConfigType:
+    ...
+
+
+@overload
+async def async_integration_yaml_config(
+    hass: HomeAssistant,
+    integration_name: str,
+    *,
+    raise_on_failure: Literal[True],
+) -> ConfigType:
+    ...
+
+
+@overload
+async def async_integration_yaml_config(
+    hass: HomeAssistant,
+    integration_name: str,
+    *,
+    raise_on_failure: Literal[False],
+) -> ConfigType | None:
+    ...
+
+
+async def async_integration_yaml_config(
+    hass: HomeAssistant, integration_name: str, *, raise_on_failure: bool = True
 ) -> ConfigType | None:
     """Fetch the latest yaml configuration for an integration."""
     integration = await async_get_integration(hass, integration_name)
-
-    return await conf_util.async_process_component_config(
-        hass, await conf_util.async_hass_config_yaml(hass), integration
-    )
+    try:
+        return await conf_util.async_process_component_config(
+            hass,
+            await conf_util.async_hass_config_yaml(hass),
+            integration,
+            raise_on_failure=raise_on_failure,
+        )
+    except ConfigValidationError as err:
+        raise ServiceValidationError(
+            str(err),
+            translation_placeholders=err.translation_placeholders,
+            translation_key="reload_config_validation_error",
+            translation_domain="homeassistant",
+        ) from err
```

Comment:
We re-raise a `ConfigValidationError` to a ServiceValidation error when it happens during a reload, and add a translation key. This allows us to have different error message on different use cases.

Reply:
I think this is weird, there's nothing saying `async_integration_yaml_config` can only be called from service handlers.
Also, why can't the reload service handlers raise `ConfigValidationError`?

Reply:
Moved catching the exception to the service handler, to be used when `raise_on_failure` is `True`.